File: May-18/05-05-18/python/01-Monk-and-the-island.py

N = 10**5 + 5
adj = {}
visit = [0]*N
level = [0]*N

def bfs(s, n):
    q = []
    q.append(s)
    visit[s] = 1
    level[s] = 0
    
    while(len(q) > 0):
        val = q[0]
        del q[0]
        if(val == n):
            return level[val]
        for i in adj[val]:
            if visit[i] == 0:
                visit[i] = 1
                level[i] = level[val] + 1
                q.append(i)
    return 0
    


for _ in range(int(input())):
    n, e = map(int, input().split())
    
    for i in range(0, n + 1):
        adj[i] = []
        visit[i] = 0
        level[i] = 0
    
    for i in range(0, e):
        x, y = map(int, input().split())
        tmp = []
        if x in adj:
            tmp = adj[x]
        tmp.append(y)
        adj[x] = tmp;
        
        tmp = []
        if y in  adj:
            tmp = adj[y]
        tmp.append(x)
        adj[y] = tmp

    print( bfs(1, n) )

File: May-18/05-05-18/java/01-Monk-and-the-island.java

import java.util.*;

class TestClass {
    static int N = (int)1e5;
    static int level[] = new int[N];
    static boolean visit[] = new boolean[N];
    static LinkedList<Integer> adj[] = new LinkedList[N];
    
    static int bfs(int s, int n)
    {
        LinkedList<Integer> q = new LinkedList<Integer>();
        q.add(s);
        visit[s] = true;
        level[s] = 0;
        
        while(q.size() != 0)
        {
            int v = q.poll();
            if(v == n)
                return level[v];
            
            Iterator<Integer> i = adj[v].listIterator();
            while (i.hasNext())
            {
                int val = i.next();
                if (visit[val] == false)
                {
                    visit[val] = true;
                    q.add(val);
                    level[val] = level[v] + 1;
                }
            }
        }
        return 0;
    }
    public static void main(String args[] ) throws Exception {
        Scanner input = new Scanner(System.in);
        int t = input.nextInt();
        while(t-- >  0)
        {
            int n, e;
            n = input.nextInt();
            e = input.nextInt();
            for(int i = 0; i <= n; i++)
            {
                adj[i] = new LinkedList();
                visit[i] = false;
                level[i] = 0;
            }
            
            for(int i = 0; i < e; i++)
            {
                int x, y;
                x = input.nextInt();
                y = input.nextInt();
                adj[x].add(y);
                adj[y].add(x);    
            }
            
            System.out.println( bfs(1, n) );
        }
        
        
        
    }

}
